### PR TITLE
feat(sns): add delivery policy to sns subscriptions

### LINKED ISSUE #17576: (sns): Allow delivery retry policy configuration of HTTPS subscription on SNS topic
### Description

While creating HTTPS Subscription on SNS topic, there's no way to configure the retry policy.

### Use Case

I am creating SNS topic and HTTPS subscription using CDK and want to provide custom delivery retry policy.

### Proposed Solution

Update UrlSubscriptionProps interface to provide deliveryRetryPolicy as well.

### Other information

_No response_

### Acknowledge

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by LaurenceWarne**
### Issue # (if applicable)

Closes #17576.

### Reason for this change

To allow setting of delivery policy for appropriate sns subscriptions.

### Description of changes

Subscriptions can now take a new parameter `deliveryPolicy` which consists of a `healthyRetryPolicy`, a `throttlePolicy`, and a `requestPolicy` (each having the same parameters described here: https://docs.aws.amazon.com/sns/latest/dg/sns-message-delivery-retries.html#creating-delivery-policy).

### Description of how you validated changes

New integration test along with unit tests.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts
Comment (LaurenceWarne) [inline packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts:5]: Not the best doc I've ever written :sweat_smile: ,  the best indicator of to what these each of these do that I found was an image at the very bottom of the doc page here: https://docs.aws.amazon.com/sns/latest/dg/sns-message-delivery-retries.html - suggestions welcome.

File: packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts
Comment (LaurenceWarne) [inline packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts:46]: Only, this and the next two appear to be mandatory if `healthyRetryPolicy` is specified, thought I can't find any specific doc for this.

**Comment by pahud** [comment]
Hi

This PR has been pending for community review for a while. Please consider posting it on the #contributing channel in cdk.dev. Community members will be on the lookout there as well for possible reviews.

Check [How to get your P2 PR community reviewed](https://github.com/aws/aws-cdk/wiki/CDK-Community-PR-Reviews#how-to-get-your-p2-pr-community-reviewed) for more details.

**Comment by LaurenceWarne** [comment]
Thanks, I've posted on the channel.

File: packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts
Comment (badmintoncryer) [reply packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts:46]: Would it be correct to assume that if these are not set, it will result in an error during deployment? If so, I am curious when the default values would be set.

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (badmintoncryer) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:156]: In order to support Duration, please make the following modifications.

```suggestion
        const minDelayTarget = healthyRetryPolicy.minDelayTarget
        const maxDelayTarget = healthyRetryPolicy.maxDelayTarget
        if (minDelayTarget.toMilliseconds() < Duration.seconds(1).toMilliseconds()) {
          throw new Error(`minDelayTarget must be at least 1 second, got ${minDelayTarget.toMilliseconds()}ms.`);
        }
        if (maxDelayTarget.toMilliseconds() < Duration.seconds(1).toMilliseconds()) {
          throw new Error(`maxDelayTarget must be at least 1 second, got ${maxDelayTarget.toMilliseconds()}ms.`);
        }
        if (minDelayTarget.toSeconds() < 1 || minDelayTarget.toSeconds() > delayTargetLimit) {
          throw new Error(`minDelayTarget must be between 1 and ${delayTargetLimit} seconds inclusive`);
        }
```

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (badmintoncryer) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:162]: Could you also adjust these if statements to support Duration?

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (badmintoncryer) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:164]: Is it necessary to validate that numRetries is an Integer?

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (badmintoncryer) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:147]: How about performing these validations inside the `renderDeliveryPolicy()`?

**Review by badmintoncryer — CHANGES_REQUESTED** [review]
Thank you for creating such a wonderful PR! I’ve quickly reviewed it!

**Comment by LaurenceWarne** [comment]
> Thank you for creating such a wonderful PR! I’ve quickly reviewed it!

Thanks for the review!  Hopefully I've addressed everything

File: packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts
Comment (LaurenceWarne) [reply packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts]: If I recall correctly from testing, it's required if and only if `healthyRetryPolicy` itself is specified, and so the default values would be set in the case `healthyRetryPolicy` is not specified (perhaps it's confusing to mention a default in the doc :thinking:)

File: packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts
Comment (LaurenceWarne) [reply packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts]: > I would appreciate it if you could consult with the maintainer again.

As in a CDK maintainer?

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (LaurenceWarne) [reply packages/aws-cdk-lib/aws-sns/lib/subscription.ts]: Yeah good catch

File: packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts
Comment (badmintoncryer) [reply packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts]: Yes! After my approval, you need to get approval from the CDK maintainer. I would appreciate it if you could kindly check it at that time!

File: packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts
Comment (badmintoncryer) [reply packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts]: How about making these two props optional and explicitly setting the default values documented within the L2 construct? These properties may be required in CloudFormation, but I felt there is no need to make them mandatory in the L2 construct.

```ts
  readonly minDelayTarget?: Duration;
  readonly maxDelayTarget?: Duration;
```

```ts
const minDelayTarget = healthyRetryPolicy.minDelayTarget ?? Duration.seconds(20);
const maxDelayTarget = healthyRetryPolicy.maxDelayTarget ?? Duration.seconds(20);
```

File: packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts
Comment (badmintoncryer) [inline packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts:1]: Could the contents of `delivery-policy.ts` be used elsewhere? If it’s only going to be used in `subscription.ts`, I thought it might be better to include everything in `subscriptions.ts`.

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (badmintoncryer) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:206]: Cloud you please show the props value like below?

```ts
        throw new Error(`numRetries must be between 0 and ${numRetriesLimit} inclusive, got: ${healthyRetryPolicy.numRetries}`);
```

**Review by badmintoncryer — CHANGES_REQUESTED** [review]
I Thank you for the revisions. I’ve added a few more comments. Once you address those, I plan to approve.

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (badmintoncryer) [reply packages/aws-cdk-lib/aws-sns/lib/subscription.ts]: Has the integer check been added?

Comment (LaurenceWarne) [reply packages/aws-cdk-lib/aws-sns/lib/subscription.ts]: Sorry, forgot to push a commit :sweat_smile:

File: packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts
Comment (LaurenceWarne) [reply packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts]: Ah right np, I thought you might have meant someone in the SNS team or something :+1:

Comment (badmintoncryer) [reply packages/aws-cdk-lib/aws-sns/lib/delivery-policy.ts]: Ref: https://github.com/aws/aws-cdk/wiki/CDK-Community-PR-Reviews

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (badmintoncryer) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:227]:
```suggestion
        minDelayTarget: (healthyRetryPolicy.minDelayTarget === undefined) ? 20 : healthyRetryPolicy.minDelayTarget.toSeconds(),
        maxDelayTarget: (healthyRetryPolicy.maxDelayTarget === undefined) ? 20 : healthyRetryPolicy.maxDelayTarget.toSeconds(),
        numRetries: (healthyRetryPolicy.numRetries === undefined) ? 3: healthyRetryPolicy.numRetries,
```

**Review by badmintoncryer — APPROVED** [review]
Thank you for you contribution! I've approved this, but I kindly request a minor correction.

**Comment by LaurenceWarne** [comment]
> Thank you for you contribution! I've approved this, but I kindly request a minor correction.

:+1: And thanks again, I really appreciate you taking the time :slightly_smiling_face:

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:169]: nit: 
```suggestion
    if (![SubscriptionProtocol.HTTP, SubscriptionProtocol.HTTPS].includes(protocol)) {
```

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:178]: I think you can directly use `toSeconds()` which will check if it's a integer second

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:199]: It's good intention to not use the magic value and define it explicitly. But given the number of validations, I actually think it's more readable to directly use the value like 100 and 3600 in the if condition checks. What do you think?

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:222]: can you explain what you meant by `empirically mandatory)`? I do see that they do have default values set.

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-sns/lib/subscription.ts:230]: Should these have default values?

**Review by GavinZZ — DISMISSED** [review]
Thanks for contributing! Left some feedback.

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (LaurenceWarne) [reply packages/aws-cdk-lib/aws-sns/lib/subscription.ts:178]: I thought about it, but I thought this would be better as the error `toSeconds` raises doesn't have the same context we can add here?

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (LaurenceWarne) [reply packages/aws-cdk-lib/aws-sns/lib/subscription.ts:199]: I think I prefer it how it is now, I agree it might be more readable, but I'd prefer keeping hard-coding to a minimum to make refactors safer/simplier, etc

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (LaurenceWarne) [reply packages/aws-cdk-lib/aws-sns/lib/subscription.ts:222]: Does https://github.com/aws/aws-cdk/pull/30830#discussion_r1770405277 answer your question (also see last comment in that thread as well as to why they're optional)?

File: packages/aws-cdk-lib/aws-sns/lib/subscription.ts
Comment (LaurenceWarne) [reply packages/aws-cdk-lib/aws-sns/lib/subscription.ts:230]: Similar to https://github.com/aws/aws-cdk/pull/30830#discussion_r1796169589 question, I guess it would depend on your view on that :sweat_smile:?

My thinking was that if Cloudformation doesn't require these to be specified when `healthyRetryPolicy` is specified, we can omit them here and have Cloudformation figure the defaults for us?

**Review by GavinZZ — APPROVED** [review]

**Comment by badmintoncryer** [comment]
@mergify update

**Comment by adoblas** [comment]
@GavinZZ do we have an ETA for this merge?

**Comment by badmintoncryer** [comment]
@GavinZZ Could you please merge main branch and approve this PR again?

**Review by GavinZZ — APPROVED** [review]

**Comment by mwebber** [comment]
@LaurenceWarne , @GavinZZ it looks like the reason this PR hasn't proceeded to the next is that we have 2 code coverage test failures (see https://app.codecov.io/gh/aws/aws-cdk/pull/30830/indirect-changes). It looks like the 2 failures are a single problem. 

The failures just look like some strange side-effect of the change, not an actual problem with the PR. However, I don't think the system will let you move forward until they are addressed.

Is someone able to resolve those failures? Sorry I have no idea how to fix them myself.

**Comment by LaurenceWarne** [comment]
> The failures just look like some strange side-effect of the change, not an actual problem with the PR. However, I don't think the system will let you move forward until they are addressed.
> 
> Is someone able to resolve those failures? Sorry I have no idea how to fix them myself.

Merged master and it looks to have fixed itself

**Comment by GavinZZ** [comment]
Yes, apologize that I didn't get back to you sooner. We've fixed issue with codecov now. Will approve and merge again.

**Review by GavinZZ — APPROVED** [review]

**Pull request merged**
